Minor doc improve (#985)
FluxML/Flux.jl: docs/src/models/basics.md

## Taking Gradients

Flux's core feature is taking gradients of Julia code. The [`gradient`](@ref) function takes another Julia function `f` and a set of arguments, and returns the gradient with respect to each argument. (It's a good idea to try pasting these examples in the Julia terminal.)

```jldoctest basics
julia> using Flux

julia> import Flux: gradient

julia> f(x) = 3x^2 + 2x + 1;

julia> df(x) = gradient(f, x; nest = true)[1]; # df/dx = 6x + 2

julia> df(2)

> 14.0 (tracked)

Reviewer: variables are not tracked since the move from Tracker to Zygote for AD

Author: I'm on Flux#master. Running the code gives this. Apparently, Zygote isn't used by default.

Author: I added `using Zygote` and the issue is gone now.

Author: @CarloLucibello my Flux had been downgraded automatically because old compat bounds. Adding #master again fixes the problem

6
```

> When a function has many parameters, we can get gradients of each one at the same time by passing vectors of corresponding inputs for each argument to `gradient`:

Reviewer 2: ```suggestion
When a function has many parameters, we can get gradients of each one at the same time by passing each one as an argument:
```

```jldoctest basics
julia> f(x, y) = sum((x .- y).^2);
```

> Now we calculate gradient of `f` for `x=2, y=2` and `x=1, y=0`:

Reviewer 2: This is not what the example is intended to show; the point is that you can get `x` and `y` at the same time, not that you can get gradients at multiple points at the same time (which is true only in very specific cases).

Author: Could you edit my sentence to show what you are trying to say, I still don't understand this part of doc. It is not clear.

Reviewer 2: Maybe it would be best to skip these parts then and just do the definite improvements; we can look at this later.

Reviewer 3: The `f` in that example is simply the square of the [Euclidean distance](https://en.wikipedia.org/wiki/Euclidean_distance) between two vectors `x ≡ (x1, x2, ..., xN)` and `y ≡ (y1, y2, ..., yN)`. So the arguments being passed are the `x ≡ (2, 1)` and `y ≡ (2, 0)` points.

So I would word it this way:
```
For this `f`, we can get the gradient with respect to both `x` and `y` in a single call:
<The gradient example here>
where `x = [2, 1]` and `y = [2, 0]`
```
Sorry about butting into the conversation. I'm trying out Julia / Flux and found a bunch of minor mistakes / inconsistencies in the documentation that I would like to see fixed too.

Reviewer 2: A wording like this sounds good to me. Perhaps if @aminya wants to revert the current version you could put up a separate PR. It's also possible to PR against this PR but that's a little complicated.

Reviewer 2: A wording like this sounds good to me. Perhaps if @aminya wants to revert the current version you could put up a separate PR. It's also possible to PR against this PR but that's a little complicated.

Author: I can commit it to this PR. no need for a separate PR.

6
```

> When a function has many parameters, we can get gradients of each one at the same time by passing each one as an argument:

Reviewer: I would keep only this change and leave the rest as it is since it seems redundant

Author: What changes are redundant? Do you mean the examples!?

Reviewer: all changes except the line above, in my opinion. I prefer the previous version

Author: I am describing the syntax. I don't dee this as redundant.

> When a function has many parameters, we can get gradients of each one at the same time by passing each one as an argument:

Reviewer 4: ```suggestion
When a function has many parameters, we can get the gradient with respect to each in a single call:
```

```jldoctest basics
julia> f(x, y) = sum((x .- y).^2);
```

> For this `f`, we can get the gradient with respect to both `x` and `y` in a single call:

Reviewer 4: ```suggestion
```
If the "with respect to" is moved up top, this line is no longer necessary.

```jldoctest basics
julia> gradient(f, [2, 1], [2, 0])
([0, 2], [0, -2])
```

> where `x = [2, 1]` and `y = [2, 0]`. The above syntax is the same as: `df(x,y) = gradient(f,x,y); df([2,1], [2,0])`

Reviewer 4: ```suggestion
The above syntax is the same as: `df(x,y) = gradient(f,x,y); df([2,1], [2,0])`
```

Reviewer 4: Having a statement like "where ..." is a little cumbersome. Since `df(x, y)` already names the parameters, there's no need to define them in prose. Alternatively, you could actually define them in the doctest with `x, y = ...`, but that might be confusing given `f(x, y)` has already been defined earlier (users who aren't comfortable with scoping may conflate the two).

julia> y = [2, 0];

> julia> f(x, y) = sum((x .- y).^2);

Reviewer 2: Is it necessary to repeat this definition here, after it's defined right above?

Author: No, because the example starts with `using Flux` again, it implies that it is a new example

Reviewer 2: Let's remove the `using Flux` instead, in that case.
